[NTOS:MM] Implement support for allocating/freeing mapping memory addresses and other stuff

oleg-dubinskiy opened the pull request:
## Purpose

Implement some MM routines for allocating/freeing mapping memory addresses, probing/locking memory pages and reserved mapping stuff.

This fixes a lot of issues:
1. Microsoft NTFS driver is now detecting NTFS partitions correctly (with also PR #3400 applied).
2. The system is now booting into desktop successfully with networking drivers from Windows XP/2003. In particular, it boots with afd.sys & tdi.sys, and the situation is much more improved with also ndis.sys replaced.
3. The atapi.sys driver from Windows XP/2003 should now also load properly in ReactOS (at 1st setup stage/ in LiveCD).

Patch by @ThFabba.

JIRA issue: [CORE-10147](https://jira.reactos.org/browse/CORE-10147), [CORE-17401](https://jira.reactos.org/browse/CORE-17401), [CORE-17409](https://jira.reactos.org/browse/CORE-17409).

## Proposed changes

Implement the following functions:

- MmMapLockedPagesWithReservedMapping;
- MmUnmapReservedMapping;
- MmProbeAndLockProcessPages;
- MmAllocateMappingAddress;
- MmFreeMappingAddress.

## TODO

- [x] Perhaps improve commit message. Since I'm not familiar with this stuff so far, I don't know exact purpose of those functions.

## Result

NTFS driver from Windows XP/2003:

![MS_ntfs_in_ReactOS_wo_USB](https://user-images.githubusercontent.com/26385117/117692717-f7437900-b1c5-11eb-8930-10ba69ee5ba0.png)

AFD and TDI drivers from Windows XP/2003:

![afd+tdi](https://user-images.githubusercontent.com/26385117/117692819-104c2a00-b1c6-11eb-9fb1-50d8560c98e7.png)

oleg-dubinskiy commented:
> Well, before reviewing, I'd like to have a word from @ThFabba about why he didn't push this to master or make this PR himself.

I asked him can I send a PR with that patch many time ago, but he didn't response yet.
The reason for such long delay might be it doesn't completely fix the afd/ndis/tdi issue (although it fixes ntfs and pciidex ones), but it's just my guessing.

binarymaster commented:
Re commit message, multiple tickets can be written like that without punctuation:
```diff
-CORE-10147, CORE-17401, CORE-17409.
+CORE-10147 CORE-17401 CORE-17409
```

oleg-dubinskiy commented:
> Mistake in the commit message: pciide => atapi

> Re commit message, multiple tickets can be written like that without punctuation:
> 
> ```diff
> -CORE-10147, CORE-17401, CORE-17409.
> +CORE-10147 CORE-17401 CORE-17409
> ```

Fixed, thanks.
PR description will be also updated accordingly. :slightly_smiling_face:

zefklop reviewed (changes requested):
MmUnmapReservedMapping needs to be fully implemented.

tkreuzer commented on ntoskrnl/mm/ARM3/pool.c:
```diff
@@ -1395,26 +1395,101 @@ MiInitializeSessionPool(VOID)
+    MMPTE TempPte;
+
+    /* How many PTEs does the caller want? */
+    SizeInPages = BYTES_TO_PAGES(NumberOfBytes);
+    if (SizeInPages == 0)
+    {
+        KeBugCheckEx(SYSTEM_PTE_MISUSE,
+                     0x100, /* Requested 0 mappings */
+                     SizeInPages,
+                     PoolTag,
+                     (ULONG_PTR)_ReturnAddress());
+    }
+
+    /* We need two extra PTEs to store size and pool tag in */
+    SizeInPages += 2;
```
This can't overflow as both `NumberOfBytes` and `SizeInPages` are `SIZE_T` and the latter is obviously a fraction of the former.

GeoB99 commented:
I'm going to close this since the author has not given any signs of activity on this PR. Feel free to continue work on this whenever you want and I'll open the PR back.

HBelusca commented:
Any news on this PR @oleg-dubinskiy ?

JoachimHenze commented on ntoskrnl/mm/ARM3/mdlsup.c:
```diff
@@ -1626,29 +1626,296 @@ MmAdvanceMdl(IN PMDL Mdl,
+    Mdl->MdlFlags &= ~(MDL_MAPPED_TO_SYSTEM_VA |
+                       MDL_PARTIAL_HAS_BEEN_MAPPED |
+                       MDL_FREE_EXTRA_PTES);
 }
```
Why? The PR has a lot of justification to be merged as is, in case it allows to boot to desktop with MS NTFS.
Why do we have to create ReactOS 1.0 here?
